### ISSUE TITLE
declare bn inside forward

### ISSUE THREAD

**Issue by lzrobots** [mapillary/inplace_abn#101]
For some reason, I need to declare the `bn` inside of  forward. There is no problem of `nn.BatchNorm2d()` in following code but when replace the `nn.BatchNorm2d()` with `InPlaceABNSync` then the programme is stuck there without printing error. Do you have any idea to make it work?

```
    def forward(self, x, b):
        if self.theta is None:
            _, _, h, w = b.size()
            self.theta = nn.Conv2d(self.h_planes, self.h_planes // self.ratio, kernel_size=1, bias=False).cuda()
            self.bn = nn.BatchNorm2d(h * w).cuda()
```

**Comment by rotabulo** [comment]
@lzrobots does it work the version without sync? By the way, by not declaring the module in the init, you probably are not optimizing the parameters of the addtional layers ... does it make sense to you?

**Comment by lzrobots** [comment]
@rotabulo Yes it works with `nn.BatchNorm2d` but doesn't work for `InPlaceABNSync`. Is there any solution to declare `bn` and `conv` that relate to the spatial size (`h, w`) of input in init?

**Comment by rotabulo** [comment]
@lzrobots I meant, did you try with `InPlaceABN` _non_sync ? 
Regarding you question, depends how you want to tie the parameters to the different input sizes.
